library bundle eligibility check T235262

jsnshrmn opened the pull request:
As talked about in the description of:
https://phabricator.wikimedia.org/T235262

In addition to adding editor object attributes for bundle eligibility, this PR breaks up the individual requirements for an account as specified in [our terms](https://wikipedialibrary.wmflabs.org/terms/) into separate attrs as well.

There are lots of other places where we're using props instead of attrs in order to keep our db schema as clean as we can. In this case, I thought attrs were better because:

- we can directly work against editor querysets that are filtered on elements of eligibility
- the data used to determine eligibility only changes upon user oauth login. Nothing the user does after login will impact the freshness of these attributes

Note that this PR doesn't currently include a cron task to update the recent edit count for editors. That means that if a user is eligible upon a given login, makes 10 edits, and then waits 60 days before logging in again, that they would be flagged as eligible. We'll want that cron task, but I thought it would be best for us to sort out the way the checks worked and how things were stored in the db first.

This PR also doesn't include any kind of display for the added fields. We'll probably want those.

Samwalton9 commented on TWLight/users/models.py:
```diff
@@ -247,46 +362,29 @@ def get_wp_groups_display(self):
+                    self.wp_enough_recent_edits = False
+        else:
+            self.wp_editcount_prev_date_updated = date.today()
+
```
Don't we want to always be updating `wp_enough_recent_edits` on login so that we can check the most up-to-date edit count? In that way an editor who doesn't qualify can go make enough edits, then come back and upon logging in be flagged as eligible.

jsnshrmn replied:
This PR is still a work in progress and isn't ready for review. It's just now to the point where it doesn't break existing code. This block is going to check either 30 days or the required delta in edit count

Samwalton9 replied:
Sure! Was just browsing through and wanted to make sure this behaviour didn't become too difficult to disentangle down the line :)

Samwalton9 commented on TWLight/users/models.py:
```diff
@@ -105,6 +105,112 @@ def create_user_profile(sender, instance, created, **kwargs):
+    return int(global_userinfo["editcount"]) >= 500
+
+
+def editor_not_blocked(identity):
```
Oh I see we actually were checking on this before, but I don't see that we were presenting that data anywhere, meaning a user could be flagged 'not eligible', but there not be a clear reason why!

jsnshrmn replied:
Yeah, I'll just add that (along with the eligibility stuff we've added) to this pr.

jsnshrmn replied:
Okay, I took a stab at adding granular info to the editor detail page. I've erased identifiers in this image.

![bundle_eligibility](https://user-images.githubusercontent.com/2986893/73086925-3184ec00-3e97-11ea-8d9b-b76c96ae3a46.png)

jsnshrmn replied:
just noticed that `satisfies recent editcount` didn't break over to a new row. fixed.

jsnshrmn commented on TWLight/users/models.py:
```diff
@@ -105,6 +105,112 @@ def create_user_profile(sender, instance, created, **kwargs):
+            ).date()
+        except (TypeError, ValueError):
+            reg_date = None
+            pass
```
Nope, just a leftover from refactor

jsnshrmn commented on TWLight/users/models.py:
```diff
@@ -105,6 +105,112 @@ def create_user_profile(sender, instance, created, **kwargs):
+    if enough_edits and account_old_enough and not_blocked:
+        return True
+    else:
+        logger.info("Editor {username} was not valid.".format(username=username))
```
Nope, just a leftover from refactor. I'm guessing this was originally dropped in for testing the validity checks.

jsnshrmn replied:
We're actually doing quite a bit of logging for the login process, and this happens to fall in the middle of it, here's a sample. Just noting it here for posterity.
```
Found get parameters for post-login redirection.
handshaker gotten.
Localizing oauth handshake URL.
handshaker initiated.
Authenticating user...
Identifying user...
User has been identified; getting or creating user.
Attempting to update editor after OAuth login.
Fetched global_userinfo for User.
Editor xxx was not valid.
Editor updated.
User has been updated.
```
My guess is that this was added during one of our many bouts of OAuth flakiness. I'm still going to remove this particular bit, because I agree that it doesn't seem particularly diagnostic.

jsnshrmn commented:
I see that I'm recording the dates for the previous edit count improperly, will fix.

Samwalton9 commented:
Very nice! The previous edit count + explanation might not be necessary, in some respects it's really a background-computation number more than one that users will be interested in/fully understand. Besides that this is looking good :)

jsnshrmn commented:
okay, I removed the previous editcount and accompanying text from the editor profile template. I've also fixed the date handling issues that I discovered relating to calculating the recent editcount. Per our off-thread conversations, we're not going to merge this to directly to master. Rather, I've force pushed this branch to staging, against which we will do dependent bundle-related PRs. Once all bundle related work is merged into staging and happy, we'll do a staging -> master PR for all bundle functionality.
